### DISCUSSION TITLE
'Closeable' return value ignored

### DISCUSSION THREAD

**Discussion by cpurdy** [xtclang/xvm#24] [Ideas]
Gene and I discussed having a compiler warning for methods returning a `Closeable` value that is ignored by the caller. This seemed reasonable at first, but having implemented it, it turns out that we have many, many places that naturally return a `Closeable` value that is ignored. For example, various buffers (supporting a fluent style interface) are both `Closeable` and typically `return this`. So `buf.add(val);` would generate this warning, and we have that in thousands of places.

Thoughts?

**Comment by markandrewfalco** [comment]
Regarding the buffer example, I imagine the reason it is ok to ignore is because the closeable is already handled in some outer scope?

**Reply by cpurdy** [reply]
The challenge is that we don't know (at compile time) what instance is being returned; we only know the type. So reporting a warning based on the type causes a billion false positives :(

**Reply by markandrewfalco** [reply]
Perhaps we need an annotation to say that a return value cannot be ignored

**Reply by cpurdy** [reply]
Instead of just looking for Closeable? i.e. "you should warn any time that someone calls this method and doesn't capture this return value"?

**Reply by ggleyzer** [reply]
It appears that we need the opposite: "don't warn if someone calls this method and doesn't capture..."

**Reply by markandrewfalco** [reply]
I think there will be far too many places where things which happen to implement Closeable are returned for opting-out to be practical, I think you have to opt-in. This really becomes a transfer of ownership annotation saying that the receiver is taking over the responsibility for the eventual close. This same annotation could be applied to parameters too, thus indicating to a caller that they are giving up that responsibility. @TransferOwnership, or perhaps @Adopt.
